include/bringauto/async_function_execution/AsyncFunctionExecutor.hpp

```diff
@@ -155,35 +207,40 @@ class AsyncFunctionExecutor {
 	 * @param function The function definition of which function to call.
 	 * @param args The arguments to pass to the function.
 	 * @return The return value of the function. If the return type contains some byte buffer,
-	 * the memory is valid until the next call to callFunc().
+	 * the memory is valid until the next call to callFunc(). On error, returns a CallError enum value.
 	 */
 	template <typename Ret, typename... FArgs, typename... CallArgs>
-	Ret callFunc(const FunctionDefinition<Ret, FArgs...> &function, CallArgs&&... args) {
+	auto callFunc(const FunctionDefinition<Ret, FArgs...> &function, CallArgs&&... args) -> std::expected<Ret, CallError> {
 		if (!settings_.isProducer) {
-			throw std::runtime_error("Cannot call function in consumer mode");
+			return std::unexpected(CallError::InvalidExecutorType);
 		}
 		if (sizeof...(CallArgs) < sizeof...(FArgs)) {
-			throw std::invalid_argument("Argument count mismatch");
+			return std::unexpected(CallError::ArgumentCountMismatch);
 		}
 		if (!isFunctionDefined(function.id)) {
-			throw std::runtime_error("Function ID not defined");
+			return std::unexpected(CallError::FunctionIdNotDefined);
 		}
-
+		if (callInProgress_[function.id.value].load()) {
+			return std::unexpected(CallError::FunctionCallInProgress);
+		}
+		
+		callInProgress_[function.id.value] = true;
 		auto messageBytes = serializeArgs(function.id, args...);
 		client_->sendMessage(function.id.value, messageBytes);
 
-		auto responseBytes = client_->waitForMessage(function.id.value + MESSAGE_RETURN_CHANNEL_OFFSET);
+		auto responseBytes = client_->waitForMessage(function.id.value + MESSAGE_RETURN_CHANNEL_OFFSET,
+													 settings_.functionConfigs.configs.contains(function.id.value) ?
+													 settings_.functionConfigs.configs[function.id.value].timeout :
```

Comment:
This looks dangerous too. Order of argument evaluation is/can_be random so this can break the program.